storage: periodically attempt to collocate raft leadership and leases

andreimatei opened the pull request:
We used to only attempt to do it when a lease is transferred, and if
that failed we never tried again. Now we (arbitrarily) retry on the
periodic lease extensions.

cc @tschottdorf @bdarnell @arjunravinarayan

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8834)

<!-- Reviewable:end -->

andreimatei committed:
storage: periodically attempt to collocate raft leadership and leases

We used to only attempt to do it when a lease is transferred, and if
that failed we never tried again. Now we (arbitrarily) retry on the
periodic lease extensions.

andreimatei commented:
Found this when trolling the code with @tschottdorf

tamird commented:
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[storage/replica_trigger.go, line 294 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834#-KQ2Hwf6a-7EdAduY92B:-KQ2Hwf7lzsgmmdyzDra:bvqrld3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8f433eb8a01aa597f5d436d1c354d5bf4a9b7cc/storage/replica_trigger.go#L294)):*
> ```Go
> ) {
> 
> 	iAmTheLeaseHolder := trigger.lease.Replica.ReplicaID == replicaID
> ```

can you think up a more idiomatic name for this?

---

*[storage/replica_trigger.go, line 296 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834#-KQ2IB97Hstziz6z3OT1:-KQ2IB97Hstziz6z3OT2:b-qd9dx5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8f433eb8a01aa597f5d436d1c354d5bf4a9b7cc/storage/replica_trigger.go#L296)):*
> ```Go
> 	iAmTheLeaseHolder := trigger.lease.Replica.ReplicaID == replicaID
> 
> 	// Is the lease changing hands and am is this replica the new lease holder?
> ```

"and am is?

---

*[storage/replica_trigger.go, line 327 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834#-KQ2I11JshMiMNKqf80b:-KQ2I11JshMiMNKqf80c:bw5cgcx) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8f433eb8a01aa597f5d436d1c354d5bf4a9b7cc/storage/replica_trigger.go#L327)):*
> ```Go
> 	}
> 
> 	// If this replica is the raft leader but it is not the new lease
> ```

this comment doesn't make sense on this line.

---

*[storage/replica_trigger.go, line 329 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834#-KQ2I6FJc5fmf75OwZY1:-KQ2I6FJc5fmf75OwZY2:b-59y2qr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8f433eb8a01aa597f5d436d1c354d5bf4a9b7cc/storage/replica_trigger.go#L329)):*
> ```Go
> 	// If this replica is the raft leader but it is not the new lease
> 	// holder, then try to transfer the raft leadership to match the
> 	// lease. We like it when leases and raft leadership are collocated.
> ```

this sentence is not useful. perhaps you could explain why such collocation is good?

---

*[storage/replica_trigger.go, line 331 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834#-KQ2IObSmNEqmJ8Uow6b:-KQ2IObSmNEqmJ8Uow6c:b-dc01lr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8f433eb8a01aa597f5d436d1c354d5bf4a9b7cc/storage/replica_trigger.go#L331)):*
> ```Go
> 	// lease. We like it when leases and raft leadership are collocated.
> 	if !trigger.lease.Covers(r.store.Clock().Now()) {
> 		return
> ```

these early returns are a bit worrying since this function is not explicitly about transferring leadership. concretely, i am concerned that there's nothing here to indicate to future contributors that they should not add logic at the end of the function.

---

*[storage/replica_trigger.go, line 342 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834#-KQ2IZqD0RNfytFYTez9:-KQ2IZqElDMdqjdt5uKw:b-90yey3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8f433eb8a01aa597f5d436d1c354d5bf4a9b7cc/storage/replica_trigger.go#L342)):*
> ```Go
> 			// behind on applying the log.
> 			//
> 			// TODO(andrei): We should do this attempt when a lease changes hands, and
> ```

Isn't this already what this change introduces? I don't understand what this TODO is about.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[storage/replica_trigger.go, line 294 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834#-KQ2Hwf6a-7EdAduY92B:-KQ2p4Mj2lZ9BqgbR25j:b-mutsun) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8f433eb8a01aa597f5d436d1c354d5bf4a9b7cc/storage/replica_trigger.go#L294)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can you think up a more idiomatic name for this?
</details>
I guess it could be `leaseHolder` although that doesn't suggest a `bool` type. But I kind of like `iAmTheLeaseHolder`. This situation is not common because one of the replicas behaves differently than others, so I think the emphasis on identity is good.

---

*[storage/replica_trigger.go, line 296 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834#-KQ2IB97Hstziz6z3OT1:-KQ2p2TRzzag6HVRqsKs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8f433eb8a01aa597f5d436d1c354d5bf4a9b7cc/storage/replica_trigger.go#L296)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "and am is?
</details>
Done.

---

*[storage/replica_trigger.go, line 327 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834#-KQ2I11JshMiMNKqf80b:-KQ2q5CD4UhCIUsvONJX:bxc1y4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8f433eb8a01aa597f5d436d1c354d5bf4a9b7cc/storage/replica_trigger.go#L327)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this comment doesn't make sense on this line.
</details>
it was supposed to be a comment for everything that comes next in the fun (similar to your concern about other people doing random stuff after the returns). Anyway, see now.

---

*[storage/replica_trigger.go, line 329 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834#-KQ2I6FJc5fmf75OwZY1:-KQ2pftzfbmSB1MQ5VV7:b-zwy5k) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8f433eb8a01aa597f5d436d1c354d5bf4a9b7cc/storage/replica_trigger.go#L329)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this sentence is not useful. perhaps you could explain why such collocation is good?
</details>
Done. (also I'm open to a refinements if you have them)

---

*[storage/replica_trigger.go, line 331 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834#-KQ2IObSmNEqmJ8Uow6b:-KQ2pb9kXnTfApwggfjg:bxr7v3g) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8f433eb8a01aa597f5d436d1c354d5bf4a9b7cc/storage/replica_trigger.go#L331)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> these early returns are a bit worrying since this function is not explicitly about transferring leadership. concretely, i am concerned that there's nothing here to indicate to future contributors that they should not add logic at the end of the function.
</details>
ok, I've moved the transfer to a separate function

---

*[storage/replica_trigger.go, line 342 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834#-KQ2IZqD0RNfytFYTez9:-KQ2pnGVbjL_Mc2hsLex:b-80yaxr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8f433eb8a01aa597f5d436d1c354d5bf4a9b7cc/storage/replica_trigger.go#L342)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Isn't this already what this change introduces? I don't understand what this TODO is about.
</details>
added a few more words, maybe they help, or maybe you have another suggestion.
The point is that we need some periodic check for the colocation, but it's arbitrary that we chose the lease extension to be our `cron`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Looks like you haven't pushed your latest changes.

This could really use some metrics - if for some reason these transfers are failing, then repeatedly failing to transfer could be causing unavailability that is worse than the latency of having the leader separate from the range lease. In fact, I believe these transfer attempts are *often* failing currently because of an upstream change we haven't picked up yet, so this change would make things worse.

So I think we should shelve this for a while, and not merge it until A) we're on the latest version of etcd (#8017) and B) we have sufficient monitoring to be able to tell how much of a difference this makes.

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834#-:-KQ4dfri1YEqmMEN2PxM:b-lixe41)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Oops, sorry, pushed now. My local branch somehow had lost its upstream.

Curious to understand - why are failed attempts to transfer particularly disruptive?
#8803 is adding a counter for the ranges with split lease/leaders. Is that good enough? Or you want a more explicit counter of failed transfer attempt? I'm not sure how to add this, that `raftGroup.TransferLeader()` looks pretty opaque, but maybe you have ideas.

Anyway I'm fine with waiting for the etcd update, I'll subscribe myself to that bug.

---

Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834#-:-KQ6lmKrJxOPocphRbdK:bwllm0n)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
> Curious to understand - why are failed attempts to transfer particularly disruptive?

I'm not familiar with the implementation, but my guess would be that a failed transfer incurs a period of unavailability of duration ~electionTimeout (think of the leader stepping down and hoping that someone else campaigns immediately, which may not happen).

bdarnell commented:
Yeah, when a transfer fails the other nodes have to wait for the election timeout to be able to reestablish leadership. 

The metrics added in #8803 cover what I'm looking for here.

bdarnell commented:
Now that we have some metrics and we're seeing a significant level of lease holders who are not raft leaders in the test clusters, it's time to move forward with this change. @andreimatei, can you rebase it on top of `master` instead of `develop`?

tschottdorf commented:
I wouldn't want to gung-ho merge this until we have some idea of why the prod clusters aren't performing well even with the recent round of changes. Even if we had all lease holders dislocated, we should be making some steady progress.

tschottdorf commented:
It appears that without the interference caused by frequent GC'ing of replicas (which has stopped on gamma after deploying https://github.com/cockroachdb/cockroach/pull/9230) we see a very high degree of colocation of lease and leader:

![screen shot 2016-09-09 at 6 21 49 pm](https://cloud.githubusercontent.com/assets/5076964/18405288/eeb92132-76be-11e6-8ab6-05e81c972f96.png)

That isn't to say that this change isn't useful, but it may be less of a priority.

petermattis commented:
The leaders w/o lease metric seems to be ticking up over time:

<img width="256" alt="screen shot 2016-09-09 at 7 58 42 pm" src="https://cloud.githubusercontent.com/assets/6201335/18406183/b0ec394e-76c7-11e6-9ba1-ba9ba0212b7f.png">

Correspondingly, MsgProp is ticking up (its the green line below):

<img width="263" alt="screen shot 2016-09-09 at 7 59 57 pm" src="https://cloud.githubusercontent.com/assets/6201335/18406201/d8a97c1c-76c7-11e6-9dbf-daf6a61e5545.png">

tschottdorf commented:
Interesting. So we're likely very good at colocating initially, but the lease is more stable than the Raft leadership. Ok, the change looks useful immediately.

bdarnell commented:
I'm not sure what caused this to change, but we're no longer seeing MsgProp and leaders-without-lease trending upwards over time. I'd still like to go ahead and get this into master fairly soon.

tschottdorf commented:
The increased range count might mean that leases now expire regularly as the frequency with which block_writer hits them has reduced. However, this is only a speculation.

---

Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834#-:-KRUwoDgjgHOLBBQk9ZP:b-dx253x)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
rebased onto master

---

Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834#-:-KRVEVvInUnzm7TXOv4g:bg6itf)*
<!-- Sent from Reviewable.io -->

tamird commented on storage/replica_trigger.go:
```diff
@@ -290,68 +290,91 @@ func (r *Replica) leasePostCommitTrigger(
-			r.mu.Lock()
-			r.mu.tsCache.SetLowWater(trigger.lease.Start)
-			r.mu.Unlock()
+	iAmTheLeaseHolder := trigger.lease.Replica.ReplicaID == replicaID
```
https://www.youtube.com/watch?v=-dJolYw8tnk

tschottdorf replied:
'Come with me if you want to lease'

On Mon, Sep 12, 2016, 17:23 Tamir Duberstein <notifications@github.com>
wrote:

> In storage/replica_trigger.go
> <https://github.com/cockroachdb/cockroach/pull/8834#discussion_r78457095>:
>
> > -			// mark of the timestamp cache. Note that clock offset scenarios are
> > -			// handled via a stasis period inherent in the lease which is documented
> > -			// in on the Lease struct.
> > -			//
> > -			// The introduction of lease transfers implies that the previous lease
> > -			// may have been shortened and we are now applying a formally overlapping
> > -			// lease (since the old lease holder has promised not to serve any more
> > -			// requests, this is kosher). This means that we don't use the old
> > -			// lease's expiration but instead use the new lease's start to initialize
> > -			// the timestamp cache low water.
> > -			log.Infof(ctx, "new range lease %s following %s [physicalTime=%s]",
> > -				trigger.lease, prevLease, r.store.Clock().PhysicalTime())
> > -			r.mu.Lock()
> > -			r.mu.tsCache.SetLowWater(trigger.lease.Start)
> > -			r.mu.Unlock()
> > +	iAmTheLeaseHolder := trigger.lease.Replica.ReplicaID == replicaID
>
> https://www.youtube.com/watch?v=-dJolYw8tnk
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/8834/files/db56ac038261f5793b707d48cce0e994f07d2a21#r78457095>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135DBOxNTEg8h9HyQZZ0UgWgHaCf8Eks5qpcIogaJpZM4JtckG>
> .
>
-- 

-- Tobias

andreimatei replied:
https://www.youtube.com/watch?v=wmin5WkOuPw

a-robinson replied:
https://www.youtube.com/watch?v=fE_H1k0mi0Q

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[storage/replica_trigger.go, line 363 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834#-KRXbFojcLA27maA-s1p:-KRXbFok57Dtjoze_NiG:b-20xdkb) ([raw file](https://github.com/cockroachdb/cockroach/blob/db56ac038261f5793b707d48cce0e994f07d2a21/storage/replica_trigger.go#L363)):*
> ```Go
> 		if raftGroup.Status().RaftState == raft.StateLeader {
> 			// Only the raft leader can attempt a leadership transfer.
> 			log.Infof(ctx, "range %v: replicaID %v transfer raft leadership to replicaID %v",
> ```

This message should probably just use `%s` on `r` instead of `r.RangeID` and `replicaID` (and then you could remove the `replicaID` parameter)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834#-:-KRXbS110vGU9XwJL8q8:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[storage/replica_trigger.go, line 363 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834#-KRXbFojcLA27maA-s1p:-KRZ11ILB3BFqyXfVmeL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/db56ac038261f5793b707d48cce0e994f07d2a21/storage/replica_trigger.go#L363)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This message should probably just use `%s` on `r` instead of `r.RangeID` and `replicaID` (and then you could remove the `replicaID` parameter)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/replica_trigger.go, line 363 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834#-KRXbFojcLA27maA-s1p:-KRZ1R4QXx4Nnb5bpq_P:bwqwo5s) ([raw file](https://github.com/cockroachdb/cockroach/blob/db56ac038261f5793b707d48cce0e994f07d2a21/storage/replica_trigger.go#L363)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> Done.
</details>
I don't think you want to log `r` at all, the context seems to already include the relevant tag.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/replica_trigger.go, line 363 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834#-KRXbFojcLA27maA-s1p:-KRZDTVBSk5-TBQuGTJL:bcu8yky) ([raw file](https://github.com/cockroachdb/cockroach/blob/db56ac038261f5793b707d48cce0e994f07d2a21/storage/replica_trigger.go#L363)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I don't think you want to log `r` at all, the context seems to already include the relevant tag.
</details>
the context has some stuff, but this message can probably use more info... I'll leave it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/replica_trigger.go, line 363 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834#-KRXbFojcLA27maA-s1p:-KRZFEEwIc79PuOeeMws:b-1f8hg7) ([raw file](https://github.com/cockroachdb/cockroach/blob/db56ac038261f5793b707d48cce0e994f07d2a21/storage/replica_trigger.go#L363)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> the context has some stuff, but this message can probably use more info... I'll leave it.
</details>
That's a nice sentiment, but it doesn't make any sense. Does this context come from the replica, or not? If it does, then you're simply duplicating the logging of the replica. cc @raduberinde

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[storage/replica_trigger.go, line 363 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834#-KRXbFojcLA27maA-s1p:-KRZIQm6qy9p3FQvrM3V:bfw4rf5) ([raw file](https://github.com/cockroachdb/cockroach/blob/db56ac038261f5793b707d48cce0e994f07d2a21/storage/replica_trigger.go#L363)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> That's a nice sentiment, but it doesn't make any sense. Does this context come from the replica, or not? If it does, then you're simply duplicating the logging of the replica. cc @raduberinde
</details>
The `replicaID` isn't part of the tag or `r.String()` so that should stay. The RangeID should be redundant with the tag (if the ctx isn't based off of Replica's ctx, `Replica.logContext(ctx)` should be used to populate the tags).

Not a big deal though, I reckon we will need to make another pass over the logs once everything settles.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8834)*
<!-- Sent from Reviewable.io -->
